### PR TITLE
Stabilize const_type_id feature

### PR DESCRIPTION
**Pull request by KodrAus**
The tracking issue for `const_type_id` points to the ill-fated #41875. So I'm re-energizing `TypeId` shenanigans by opening this one up to see if there's anything blocking us from stabilizing the constification of type ids.

Will wait for CI before pinging teams/groups.

-----

This PR stabilizes the `const_type_id` feature, which allows `TypeId::of` (and the underlying unstable intrinsic) to be called in constant contexts.

There are some [sanity tests](https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const-typeid-of-rpass.rs) that demonstrate its usage, but I’ve included some more below.

As a simple example, you could create a constant item that contains some type ids:

```rust
use std::any::TypeId;

const TYPE_IDS: [TypeId; 2] = [
    TypeId::of::<u32>(),
    TypeId::of::<i32>(),
];

assert_eq!(TypeId::of::<u32>(), TYPE_IDS[0]);
```

Type ids can also now appear in associated constants. You could create a trait that associates each type with its constant type id:

```rust
trait Any where Self: 'static {
    const TYPE_ID: TypeId = TypeId::of::<Self>();
}

impl<T: 'static> Any for T { }

assert_eq!(TypeId::of::<usize>(), usize::TYPE_ID);
```

`TypeId::of` is generic, which we saw above in the way the generic `Self` argument was used. This has some implications for const evaluation. It means we can make trait impls evaluate differently depending on information that wasn't directly passed through the trait system. This violates the _parametricity_ property, which requires all instances of a generic function to behave the same way with respect to its generic parameters. That's not unique to `TypeId::of`, other generic const functions based on compiler intrinsics like `mem::align_of` can also violate parametricity. In practice Rust doesn't really have type parametricity anyway since it monomorphizes generics into concrete functions, so violating it using type ids isn’t new.

As an example of how impls can behave differently, you could combine constant type ids with the `const_if_match` feature to dispatch calls based on the type id of the generic `Self`, rather than based on information about `Self` that was threaded through trait bounds. It's like a rough-and-ready form of specialization:

```rust
#![feature(const_if_match)]

trait Specialized where Self: 'static {
    // An associated constant that determines the function to call
    // at compile-time based on `TypeId::of::<Self>`.
    const CALL: fn(&Self) = {
        const USIZE: TypeId = TypeId::of::<usize>();

        match TypeId::of::<Self>() {
            // Use a closure for `usize` that transmutes the generic `Self` to
            // a concrete `usize` and dispatches to `Self::usize`.
            USIZE => |x| Self::usize(unsafe { &*(x as *const Self as *const usize) }),
            // For other types, dispatch to the generic `Self::default`.
            _ => Self::default,
        }
    };

    fn call(&self) {
        // Call the function we determined at compile-time
        (Self::CALL)(self)
    }
    
    fn default(x: &Self);
    fn usize(x: &usize);
}

// Implement our `Specialized` trait for any `Debug` type.
impl<T: fmt::Debug + 'static> Specialized for T {
    fn default(x: &Self) {
        println!("default: {:?}", x);
    }
    
    fn usize(x: &usize) {
        println!("usize: {:?}", x);
    }
}

// Will print "usize: 42"
Specialized::call(&42usize);

// Will print "default: ()"
Specialized::call(&());
```

Type ids have some edges that this stabilization exposes to more contexts. It's possible for type ids to collide (but this is a bug). Since they can change between compiler versions, it's never valid to cast a type id to its underlying value.

### CODE REVIEW

File: src/test/ui/consts/const-typeid-of.rs
Comment (ecstatic-morse) [inline src/test/ui/consts/const-typeid-of.rs:8]: Please make this "run-pass" and check that it works.

**Comment by ecstatic-morse** [comment]
I don't see any hazards from this. The result of `type_id` is allowed to change across compiler versions, but I think that is acceptable for `const fn`.

cc @rust-lang/wg-const-eval

**Comment by RalfJung** [comment]
Cc @rust-lang/lang because this is about const-stabilizing an intrinsic.

File: src/test/ui/consts/const-typeid-of.rs
Comment (RalfJung) [reply src/test/ui/consts/const-typeid-of.rs:8]: (Unless we already have another test that does that -- I am not sure if we do.)

Comment (KodrAus) [reply src/test/ui/consts/const-typeid-of.rs:8]: Ah we do have this one already: https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const-typeid-of-rpass.rs

**Comment by Y0ba** [comment]
What about collisions #10389? Currently this API is really dangerous to use and there is no way to detect collisions even without using dylibs. Not really related to const-stabilization but in current shape it's pretty useless both in const and in runtime.

**Comment by ecstatic-morse** [comment]
In my opinion, the fact that `type_id` is poorly implemented doesn't have any bearing on whether or not it can be `const`. All code that uses `Any` is using `type_id` under the hood, so it's not "pretty useless". While I totally agree that the chance of a hash-collision is a serious problem, saying it's "really dangerous" is an overstatement: The probability of a 64-bit hash collision in a program with 100,000 types is on the order of 1 in 10 billion.

**Comment by RalfJung** [comment]
Well, except that assigning probabilities to a deterministic function doesn't make that much sense. `type_id` collisions [have been constructed before](https://github.com/rust-lang/rust/issues/10389#issuecomment-147919570).

**Comment by ecstatic-morse** [comment]
It''s possible to intentionally find `type_id` collisions. Why should that affect const-stabilization?

**Comment by RalfJung** [comment]
I am not sure if it does. But I am arguing that your reasoning based on probabilities is meaningless.

The rest of your post I agree with.

**Comment by ecstatic-morse** [comment]
If you wish to be pedantic:

> While I totally agree that the chance of a hash-collision is a serious problem, saying it's "really dangerous" is an overstatement for the vast majority of programs: The probability of a 64-bit hash collision in a program with 100,000 types is on the order of 1 in 10 billion.

What part of this do you disagree with?

**Comment by Y0ba** [comment]
>It''s possible to intentionally find type_id collisions.

Unfortunately it's possible to hit collisions unintentionally.
> Why should that affect const-stabilization?

Because there will be more ways to hit it. That's what I wanted to say. I'd rather see that feature deprecated and switch `Any` trait to safer alternative.

**Comment by RalfJung** [comment]
> What part of this do you disagree with?

The part where you use the language of probability. Not everything is/has a probability. So by saying "the probability of X is Y", you are implicitly making a very strong claim -- that X can be meaningfully described with probabilities. That is simply not the case here.

Specifically, "The probability of a 64-bit hash collision in a program with 100,000 types" is not "on the order of 1 in 10 billion". Rather, there is no probability we can assign to this process, as it is not a probabilistic process.

Anyway, this is not leading anywhere, so I'll stop here. I'm happy to continue in some "off-topic" thread on Zulip or so.

**Comment by ecstatic-morse** [comment]
So because I failed to specify that programs that intentionally cause a `type_id` hash collision fall outside of "the vast majority of programs", you conclude that " your reasoning based on probabilities is meaningless"?

Debates should be held in good-faith, especially between rust-lang contributors. I don't feel you were trying to actually resolve the issue at hand here.

**Comment by ecstatic-morse** [comment]
> Because there will be more ways to hit it. 

This is far too conservative, IMO. It's important to me that const-eval is brought to parity with runtime evaluation when possible, and no fundamental reason has been proposed that would prevent this from being `const`. There may be a "const-safety" argument to be made here, but I think it's incorrect.

If you care about `type_id` collisions, you should bring attention to #10389, not try to stop constification withot good reason.

**Comment by RalfJung** [comment]
I'm sorry if I came across as acting in or assuming bad faith. I did neither of these things.

I assumed you maybe were not aware that probabilities are a more subtle subject than most people think. (As an example, another common, related issue is that many people conflate "non-determinism" and "randomness", and often a non-deterministic function is described as returning a "random" result. That is wrong in a similar way as your statement above.) Maybe that assumption was wrong, but it was IMO the best guess I could make (in good faith) for what motivated your original statement and its follow-ups.

I don't think I am acting in bad faith when I am pointing out that people use (precisely defined mathematical) terms the wrong way -- I think that is part of having a constructive discussion.

I would agree I am being nitpicky, but that's not the same thing as bad faith. I don't think nitpicking is necessarily a bad thing when discussing subtle (technical) details such as soundness concerns.

**Comment by RalfJung** [comment]
> not try to stop constification without good reason.

This is now the second time that you claim I want to stop constification. I do not want to stop it, and never said I did.

I just wish the arguments we are making here to be factually correct, that is all. So even if I agree with your conclusion that constification is reasonable (which I do), I still think I should point out when I do not agree with the arguments that lead you to that conclusion.

**Comment by ecstatic-morse** [comment]
https://github.com/rust-lang/rust/pull/72488#issuecomment-633289306 is meant specifically as a reply to @Y0ba.

The purpose of good-faith debate is ultimately to reach the best decision possible, not to just score points against your opponent, so to bring this back on track:

Does the fact that it is possible to *intentionally* create a `type_id` collision mean we can't make this intrinsic `const fn` given the relatively low probability of an accidental collision I stated above?

**Comment by RalfJung** [comment]
I'm not trying to score any points, I am trying to have a precise and technical discussion. Please also apply the "assume good faith" advise to yourself.

**Comment by ecstatic-morse** [comment]
I think "Good-faith" and "bad-faith" are also terms worth defining.

**Comment by Y0ba** [comment]
> It's important to me that const-eval is brought to parity with runtime evaluation when possible, and no fundamental reason has been proposed that would prevent this from being const.

Well, I just see `type_id` constification as "extending" functionality of poorly implemented feature (correctness of which depends on your luck). Not sure if bringing const-eval on par with runtime worth it 🤷.

**Comment by withoutboats** [comment]
I've marked some discussion off topic. @RalfJung describing another person's comment, which has a pretty clear colloquial interpretation, as "meaningless" is never a kind and friendly way to make an edifying side remark, as your comment about probability could have been.

@Y0ba you've used some pretty stark language to describe #10389, "really dangerous," "pretty useless," correctness depends on luck." This doesn't reflect the level of risk that I understand as embodied in #10389.

However, its not really relevant here: `TypeId`s are an opaque type, which can implement equality by any mechanism we decide. Fixing #10389 does not imply deprecating the type_id function, and I don't know that anyone has any intention to.

**Comment by Y0ba** [comment]
> This doesn't reflect the level of risk that I understand as embodied in #10389.

Since `TypeId` mostly used for dynamic type casts (as part of `Any` trait), collisions potentially could lead to transmutations between different types in safe code. I was unlucky and hit this bug in my code when tried to create type map, and after some research found #10389 (btw I think that bug should be mentioned in documentation of `Any` and `TypeId` types). After that I implemented ad-hoc solution instead of this type. That's why I call it dangerous (because it could lead to UB) and useless (because it doesn't reliably do what is was designed for).

But if `TypeId` can be changed without breaking backward compability than it's indeed not relevant here.

**Comment by ecstatic-morse** [comment]
> I was unlucky and hit this bug in my code when tried to create type map.

@Y0ba Would you mind briefly documenting your experience in the comments of #10389? I think that part of the lack of urgency on that issue is that there were no reports from users of it happening organically. Perhaps changing this would help spur action?

Also, I'm sorry I was curt with you above. I want to reiterate that I do think #10389 is a serious issue, and while I don't think that alone should preclude it from becoming `const`, I do understand your reticence around making `type_id` usable in more places.

**Comment by withoutboats** [comment]
Yea, @Y0ba, user feedback about encountering the issue in practice would probably raise the priority of resolving #10389.

**Comment by oli-obk** [comment]
I remember two arguments against stabilizing the constness of `TypeId::of`, but I can't find the discussions right now. Maybe I'm confusing the arguments with the arguments for/against `type_name`?

1. like at runtime, you can break referential transparency, but at compile-time that is worse, because you can now make generics do things differently depending on the type, without such information being exposed via the trait system. Preexisting for this is `std::mem::align_of`, but that's not as granular as `TypeId::of`
2. If you transmute the result of `TypeId::of` to an integer and use that, you have essentially created a random number generator at compile time. That's not too problematic, as the value is the same for a single compilation, so it's just like a globally seeded deterministic random number generator.

In my personal opinion neither of these are blockers, as it just becomes easier to do these kind of things, it's not like it wasn't possible before in some manner.

**Comment by ecstatic-morse** [comment]
@oli-obk, would you be willing to expand a bit on the concerns regarding referential transparency at compile-time? As I understand it, even if someone were to use `type_id` to emulate dynamic dispatch within e.g. an associated constant, `X` would still be referentially transparent if we consider `T` one of the "inputs" to `<T as Const>::X`. Perhaps my understanding of "referential transparency" is incomplete as it pertains to Rust, and I need to consider trait bounds (or the lack thereof)?

```rust
trait Const { const X: i32 }
impl<T> Const for T {
    const X: i32 = match TypeId::of::<T>() {
        // ...
    }
}
```

I am mildly concerned about 2, since stabilizing `type_id` might close off some of the potential design space around "unconst" operations for no real reason. Is it possible that we end up with a blacklist for the input/output type of `transmute` at compile-time (in this case one including `TypeId`)? OTOH, it seems to me `mem::size_of::<T>()` has basically the same set of issues here, since it's not deterministic between compiler versions either. Obviously it's much less likely to change than `type_of` though.

**Comment by oli-obk** [comment]
> would you be willing to expand a bit on the concerns regarding referential transparency at compile-time?

I'll do my best, but not that I am by no means understanding the farther ramnifications.

> As I understand it, even if someone were to use `type_id` to emulate dynamic dispatch within e.g. an associated constant, `X` would still be referentially transparent if we consider the generic parameter `T` to be one of the "inputs" to `<T as Const>::X`. Perhaps my understanding of "referential transparency" is incomplete as it pertains to Rust, and I need to consider generic parameters differently?

I don't know if my mental picture fits, but I think of it as a generic parameter without a trait bound like a function argument without a type. You can't do

```rust
fn foo<T>(x: T) {
    assert!(x == 0);
}
```

So the expectation is that you can't do

```rust
impl<T> Foo<T> {
    const _: () = assert!(TypeId::of<T>() == TypeId::of<usize>());
}
```

because there's no bound on `T` ("no type of type") that would allow you to do this check. While `size_of` and `align_of` already break this, we can even say that `size_of` doesn't really break anything, because it requires `T: Sized` and we just have that implicitly.

This is way out of the area of type system theory that I'm familiar with. I basically considered all of `type_name` and `type_id` not to be touched until  @Centril is back and I can convince them that things will be fine in practice. I can't with any confidence say what implications this has. If anyone else has the relevant knowledge and tells me it's fine, I'm fine with it.



> I am mildly concerned about 2, since stabilizing type_id might close off some of the potential design space around "unconst" operations for no real reason.

I was considering that, but I don't think there are any problems here.

> Is it possible that we end up with a blacklist for the input/output type of transmute at compile-time (in this case one including TypeId)?

No, there's no point, as that's just the obvious way. There are also raw pointer derefs and union field accesses to consider, and it becomes impossible to realistically restrain anything here. 

> OTOH, as you alluded to, mem::size_of::<T>() has basically the same set of issues here, since it's not deterministic between compiler versions either. Obviously it's much less likely to change than type_of though.

The representation/value changing is not the problem. If users opt-in to this kind of unstable behaviour, it doesn't matter whether they do it at runtime or compile-time. They need to be able to handle it. I don't think there's any problem here, but it is something I want us all to be in the clear about.

In theory we could just make `TypeId::of` return an abstract pointer or some new other abstract value that can't be inspected at compile-time just like pointers. This is not necessarily trivial to do at compile time though. We'd need to start allowing comparing abstract things to allow https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=46cc90c16f3df091a86a975960af787a

Right now you can't compare abstract pointers, not even with concrete integers, as at runtime they may be equal and at compile-time they may not necessarily be. This isn't an issue here, as the type id is just an integer and everything works out nicely. It does mean that users can transmute the type id to an int and compare against that, causing random behaviour depending on the hash algorithm and implementation choices in the compiler.

**Comment by ecstatic-morse** [comment]
> I can't with any confidence say what implications this has. If anyone else has the relevant knowledge and tells me it's fine, I'm fine with it.

Ah, the "type of type" explanation makes sense to me intuitively. Consider me squarely in this camp as well :smile:.

> In theory we could just make `TypeId::of` return an abstract pointer or some new other abstract value that can't be inspected at compile-time just like pointers.

I find this idea appealing, although I have no clue how to implement it myself. I agree that just forbidding it in `transmute` is not sufficient.

Thanks for making a "const safety" case, even if (I think) you're not quite sure where you come down on it. I also wasn't aware of the referential transparency concerns, although once again this doesn't obviously break something. Given that no one has expressed a real need for `const_type_id` (I think this PR is more a point of procedure), do we want to block this on having a better story around const safety and/or type-level compile-time referential transparency?

**Comment by oli-obk** [comment]
> I find this idea appealing, although I have no clue how to implement it myself.

Yea, I like it because it makes everything very explicit. I remember talking with @RalfJung about it and remember agreeing that it is overkill. It would be forward compatible to just doing what this PR is doing though, so if we're unsure, we can always put in the effort to implement it, there's no real downside except impl complexity.

I have some ideas for implementing it as it's the same feature as symbolic execution, but it's definitely far from trivial and not clear whether we want it at all.

**Comment by KodrAus** [comment]
> Given that no one has expressed a real need for const_type_id (I think this PR is more a point of procedure), do we want to block it on having a better story around const safety and/or type-level compile-time referential transparency?

I’m not sure I would call it a _real need_, but had a usecase in something I was tinkering with that lead here. It is effectively a map of type ids to functions that assume the input has that type:

```rust
const SPECIAL_CASES: [(TypeId, fn(&dyn Any))] = ..;
```

where the `const` really just came in as an optimisation. It’s like runtime specialization and my specific case can actually be covered by `min_specialization`. I don’t have a pressing need for this, but was keen to get the ball rolling.

**Comment by RalfJung** [comment]
FWIW, in terms of const soundness, I cannot think of a const soundness property that would be violated by `type_id`. Some of what @oli-obk mentioned above alludes to parametricity, but I would argue Rust already does not have type parametricity -- unsurprisingly, since Rust is monomorphization-based. (Rust *does* have some form of lifetime parametricity, which is related to the fact that lifetimes get erased.) `align_of` and specialization both violate parametricity.

I am not sure to what extend "type in type"-style arguments apply here. Adding `Type: Type` to a type theory "breaks" it in the sense that you can write diverging programs. However, CTFE Rust is already not a strongly normalizing type theory (well, one `loop` is stabilized). Being able to write diverging programs is something we consider acceptable. If we do, "type in type" (to my knowledge) poses no further problems.  In particular, it does not violate type safety.

But I know that @Centril had plans for const-eval that go much beyond mine, so I cannot say to what extend `type_id` would be in conflict with those plans.

> Yea, I like it because it makes everything very explicit. I remember talking with @RalfJung about it and remember agreeing that it is overkill. It would be forward compatible to just doing what this PR is doing though, so if we're unsure, we can always put in the effort to implement it, there's no real downside except impl complexity.

Given the amount of pain caused by symbolic `AllocId` I was and still am not a big fan of introducing more symbolic values to the core engine. (I have some plans for cleaning that up, hopefully I'll get to them eventually... but other stuff keeps popping up.^^)

IMO impl complexity is a huge issue in the Miri engine because having the code be as clean as possible is the best way I know to make sure the interpreter makes any sense. The Rust type system is not strong enough to help us here. Sure, we have plenty of tests, but basically each time I did some refactoring to improve clarity of the interfaces used inside Miri I found some bugs that were simply not covered by the test suite. Testing an interpreter is incredibly hard.

**Comment by withoutboats** [comment]
> If anyone else has the relevant knowledge and tells me it's fine, I'm fine with it.

Rust does not have type parametricity, it's fine.

**Comment by oli-obk** [comment]
Ok, so from my perspective, we should totally merge this. I don't think we need to explain in the docs why looking at the bits of a `TypeId` is a bad idea. If we did that, we'd have to explain it for all other types, too.

**Review by oli-obk — APPROVED** [review]

**Comment by nikomatsakis** [comment]
Before we stabilize, I believe this PR ought to have a stabilization report attached somewhere, ideally one that summarizes

* what is being stabilized
* what are some of the considerations (e.g., some of the arguments made in this thread)
* some example test cases that show the capabilities of the feature being tested

**Comment by KodrAus** [comment]
@nikomatsakis No problem! Would an update to the OP be enough?

**Comment by nikomatsakis** [comment]
@KodrAus Sure!

**Comment by KodrAus** [comment]
Alrighty, I’ve updated the OP with some notes about what’s actually being stabilized here and what you can do with it.

I may have totally misunderstood the arguments about parametricity 🙂

**Comment by nikomatsakis** [comment]
Thanks @KodrAus -- that all sounded about right to me. One thing I was hoping to get is a few pointers to test cases that demonstrate this function working as expected (and ideally some notes on what they test).

For example, it seems like it'd be good to test that the type-id we get from const evauation is, in fact, equal to the runtime one. =)

**Comment by KodrAus** [comment]
It’s not very discoverable here because it was barely touched in this PR, but we do have [some UI tests for const type ids](https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const-typeid-of-rpass.rs) that ensure they’re equivalent at compile time and runtime 🙂 I’ve added a link to the report so we can find them more easily.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

I propose that we accept the `type_id` function as eligible to run at compilation time. The stabilization report is the "head post" on the PR.

**Comment by oli-obk** [comment]
ping @cramertj @pnkfelix @scottmcm @withoutboats this FCP has been sitting idle for 3 weeks waiting for a concern or a checkmark from you

**Comment by nbdd0121** [comment]
@rodrimati1992 discovered a bug in `core::intrinsics::type_id` implementation at #73976. It needs to be fixed before this could continue.

**Comment by KodrAus** [comment]
Once #74538 is merged (waiting on bors) this should be unblocked.

**Comment by KodrAus** [comment]
Ah looks like I’ve got a test to fix up:

```

error[E0635]: unknown feature `const_type_id`
12379
  --> /checkout/src/test/ui/consts/issue-73976-monomorphic.rs:8:12
12380
   |
12381
LL | #![feature(const_type_id)]
12382
   |            ^^^^^^^^^^^^^
12383

12384
error: aborting due to previous error
```

**Comment by jplatte** [comment]
Stability attribute is 1.46, has to be updated to 1.47.

**Comment by eddyb** [comment]
Just came across this, some of the people involved in this might be interested in https://github.com/rust-lang/rust/pull/75923#issuecomment-683130518.

**Comment by scottmcm** [comment]
Interesting, eddyb.  FWIW I'd be fine with reverting this on beta under a "new information came in that made us not confident in the promise any more" justification.  Wouldn't be the first rolled-back stabilization...

**Comment by RalfJung** [comment]
See [this discussion](https://github.com/rust-lang/rust/pull/75923#issuecomment-696474511) for why stabilization was likely premature, and we should take it back before the release.

@KodrAus can you create the PR that reverts this, or should I?

**Pull request merged**
